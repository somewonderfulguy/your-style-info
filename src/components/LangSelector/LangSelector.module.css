.langSelector {
  position: relative;
  display: inline-block;
}

.langSelectorDark {
  composes: langSelector;
}

.langSelector button {
  font-size: 15px;
  color: inherit;
  cursor: pointer;
}

.langSelectorDark button {
  font-size: 15px;
  color: #919191;
  cursor: pointer;
}

.gray button {
  color: #696969;
}

.langSelectorDark.gray button {
  color: #919191;
}
.langSelectorDark.gray > button {
  color: #696969;
}

.langSelectorInner {
  display: flex;
  align-items: center;

  opacity: 0.7;
  transition: opacity 200ms;
}

.langSelectorInner:hover {
  opacity: 1;
}

.langSelectorInner:disabled {
  opacity: 0.4;
}

.icon,
.langSelectorDark.gray .icon {
  margin-right: 5px;
  fill: #696969;
}

.langSelectorDark .icon {
  fill: #919191;
}

.langSelector .triangle {
  margin-left: 5px;
  width: 0;
  height: 0;
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-top: 5px solid rgb(105, 105, 105);
  border-bottom: 0;
}

.triangleReverse {
  composes: triangle;
  transform: rotate(180deg);
}

.langSelector .dropDown {
  position: absolute;
  right: 0;
  z-index: 1;

  list-style: none;
  margin: 2px 0;
  padding: 8px 0;

  border: 1px solid rgba(var(--primaryBorder), 0.6);
  border-radius: 4px;
  background-color: rgb(var(--primaryBackground));
}

.dropDownAbove {
  composes: dropDown;
  top: 0;
  transform: translateY(calc(-100% - 8px));
}

.dropDownBelow {
  composes: dropDown;
  bottom: 0;
  transform: translateY(calc(100% + 4px));
}

.dropDown button {
  width: 100%;
  padding: 4px 24px;
  text-align: left;
  transition: background-color 0.2s;
}

.dropDown button:hover {
  background-color: rgba(var(--primaryHoverHighlight), 0.7);
}

.dropDown button:disabled {
  cursor: inherit;
  color: rgb(var(--primaryDisabled));
}

.dropDown button:disabled:hover {
  background-color: transparent;
}

/* TODO FIX ? */
.dropDown button:active {
  color: #fff;
  background-color: #007bff;
}
