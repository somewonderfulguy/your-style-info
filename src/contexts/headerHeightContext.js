import React, {createContext, useContext, useMemo, useState} from 'react'

const HeaderHeightContext = createContext()

const useHeaderHeight = () => {
  const context = useContext(HeaderHeightContext)
  if(!context) {
    throw new Error('useHeaderHeight must be used within a HeaderHeightProvider')
  }
  return context
}

const HeaderHeightProvider = props => {
  const [headerHeight, setHeaderHeight] = useState(0)
  const value = useMemo(() => ({headerHeight, setHeaderHeight}), [headerHeight])
  return <HeaderHeightContext.Provider value={value} {...props} />
}

export {HeaderHeightProvider, useHeaderHeight}