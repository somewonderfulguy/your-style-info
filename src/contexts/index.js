export * from './headerHeightContext'
export * from './loadingContext'
export * from './localisationContext'
export * from './screenDimensionsContext'
export * from './themeContext'