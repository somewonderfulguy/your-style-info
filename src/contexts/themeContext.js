import React, {createContext, useCallback, useContext, useMemo, useEffect, useState} from 'react'

const ThemeContext = createContext()

const useTheme = () => {
  const context = useContext(ThemeContext)
  if(!context) {
    throw new Error('useTheme must be used within a ThemeProvider')
  }
  const [isDarkTheme, setDarkTheme] = context

  const switchTheme = useCallback(
    () => setDarkTheme(!isDarkTheme),
    [isDarkTheme, setDarkTheme]
  )

  return {isDarkTheme, switchTheme}
}

const ThemeProvider = props => {
  const preferDarkQuery = '(prefers-color-scheme: dark)'
  const [isDarkTheme, setDarkTheme] = useState(
    () =>
      window.localStorage.getItem('isDarkTheme') === 'true' ||
      (window.matchMedia(preferDarkQuery).matches ? true : false),
  )

  useEffect(() => {
    const mediaQuery = window.matchMedia(preferDarkQuery)
    const handleChange = () => setDarkTheme(mediaQuery.matches ? true : false)
    mediaQuery.addListener(handleChange)
    return () => mediaQuery.removeListener(handleChange)
  }, [])

  useEffect(() => {
    window.localStorage.setItem('isDarkTheme', isDarkTheme)
  }, [isDarkTheme])

  const value = useMemo(() => [isDarkTheme, setDarkTheme], [isDarkTheme])

  return <ThemeContext.Provider value={value} {...props} />
}

export {ThemeProvider, useTheme}