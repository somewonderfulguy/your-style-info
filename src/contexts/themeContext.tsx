import React, {
  createContext, Dispatch, ReactNode, SetStateAction,
  useCallback, useContext, useEffect, useState
} from 'react'

type dispatchType = Dispatch<SetStateAction<boolean>>

const ThemeStateContext = createContext<boolean | undefined>(undefined)
ThemeStateContext.displayName = 'ThemeStateContext'

const ThemeDispatchContext = createContext<dispatchType | undefined>(undefined)
ThemeDispatchContext.displayName = 'ThemeDispatchContext'

const ThemeProvider = ({ children }: { children: ReactNode | ReactNode[] }) => {
  const preferDarkQuery = '(prefers-color-scheme: dark)'
  const [isDarkTheme, setDarkTheme] = useState(
    () =>
      window.localStorage.getItem('isDarkTheme') === 'true' ||
      (window.matchMedia(preferDarkQuery).matches ? true : false)
  )

  useEffect(() => {
    const mediaQuery = window.matchMedia(preferDarkQuery)
    const handleChange = () => setDarkTheme(mediaQuery.matches ? true : false)
    mediaQuery.addListener(handleChange)
    return () => mediaQuery.removeListener(handleChange)
  }, [])

  useEffect(() => {
    window.localStorage.setItem('isDarkTheme', String(isDarkTheme))
  }, [isDarkTheme])

  return (
    <ThemeStateContext.Provider value={isDarkTheme}>
      <ThemeDispatchContext.Provider value={setDarkTheme}>
        {children}
      </ThemeDispatchContext.Provider>
    </ThemeStateContext.Provider>
  )
}

const useThemeState = () => {
  const isDarkTheme = useContext(ThemeStateContext)

  if(isDarkTheme === undefined) {
    throw new Error('useThemeState must be used within a ThemeProvider')
  }

  return isDarkTheme
}

const useThemeSwitch = () => {
  const setDarkTheme = useContext(ThemeDispatchContext)

  if(setDarkTheme === undefined) {
    throw new Error('useThemeSwitch must be used within a ThemeProvider')
  }

  const switchTheme = useCallback(
    () => setDarkTheme(prevTheme => !prevTheme),
    [setDarkTheme]
  )

  return switchTheme
}

export {ThemeProvider, useThemeState, useThemeSwitch}