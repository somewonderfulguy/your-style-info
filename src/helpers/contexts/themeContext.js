import React, {createContext, useContext, useMemo, useEffect, useState} from 'react'

const ThemeContext = createContext()

const useTheme = () => {
  const context = useContext(ThemeContext)
  if(!context) {
    throw new Error('useTheme must be used within a ThemeProvider')
  }
  const [isDarkTheme, setDarkTheme] = context

  const switchTheme = () => {
    setDarkTheme(!isDarkTheme)
    localStorage.setItem('isDarkTheme', !isDarkTheme)
  }

  return {isDarkTheme, switchTheme}
}

const ThemeProvider = props => {
  const [isDarkTheme, setDarkTheme] = useState(false)
  const [hasThemeLoaded, setThemeLoaded] = useState(false)

  useEffect(() => {
    const isLocalStorageDark = localStorage.getItem('isDarkTheme')
    setDarkTheme(isLocalStorageDark === 'true')
    setThemeLoaded(true)
  }, [])

  const value = useMemo(() => [isDarkTheme, setDarkTheme], [isDarkTheme])

  if(!hasThemeLoaded) {
    // avoiding render default light theme and then re-render to
    // dark theme if localStorage keeps dark mode as true
    return <div />
  }

  return <ThemeContext.Provider value={value} {...props} />
}

export {ThemeProvider, useTheme}