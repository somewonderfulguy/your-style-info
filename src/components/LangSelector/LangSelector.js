import React, {useRef, useState} from 'react'
import {bool} from 'prop-types'

import {useLocalisation, useTheme} from 'contexts'
import {useOutsideClick} from 'shared/hooks'
import {LanguageIcon} from 'assets/images'
import {LANGUAGES, LOCALES} from 'constants/index'
import styles from './LangSelector.module.css'

const propTypes = {
  showAbove: bool,
  gray: bool
}

const defaultProps = {
  showAbove: false,
  gray: false
}

const LangSelector = ({showAbove, gray}) => {
  const {locale: currentLocale, setLocale} = useLocalisation()
  const {isDarkTheme} = useTheme()
  const langSelectorRef = useRef(null)
  const menuRef = useRef(null)
  const [isOpen, setOpen] = useState(false)

  useOutsideClick(langSelectorRef, () => setOpen(false))

  const langSelectorClass = isDarkTheme ? styles.langSelectorDark : styles.langSelector
  const grayClass = gray ? styles.gray : ''

  const triangleClass = showAbove
    ? (isOpen ? styles.triangle : styles.triangleReverse)
    : (isOpen ? styles.triangleReverse : styles.triangle)

  return (
    <div
      ref={langSelectorRef}
      className={`${langSelectorClass} ${grayClass}`}
    >
      <button
        onClick={() => setOpen(isOpen => !isOpen)}
        className={styles.langSelectorInner}
        type="button"
        // TODO switch label depending on language
        aria-label="Switch language"
        aria-expanded={isOpen}
        aria-haspopup
      >
        <LanguageIcon width={20} height={20} className={styles.icon} />
        <span>{LANGUAGES.get(currentLocale)}</span>
        <div className={triangleClass} />
      </button>

      {isOpen && (
        <ul
          ref={menuRef}
          className={showAbove ? styles.dropDownAbove : styles.dropDownBelow}
          role="menu"
        >
          {LOCALES.map(locale => (
            <li role="menuitem" key={locale}>
              <button
                type="button"
                disabled={locale === currentLocale}
                onClick={() => {
                  setLocale(locale)
                  setOpen(false)
                }}
              >
                {LANGUAGES.get(locale)}
              </button>
            </li>
          ))}
        </ul>
      )}
    </div>
  )
}

LangSelector.propTypes = propTypes
LangSelector.defaultProps = defaultProps

export default LangSelector