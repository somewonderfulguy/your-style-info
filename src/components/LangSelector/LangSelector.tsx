import { useRef, useState } from 'react'

import { useThemeState } from '~contexts/themeContext'
import { useLocalization } from '~contexts/localizationContext'
import { useOutsideClick } from '~shared/hooks'
import { LanguageIcon } from '~assets/images'
import { LANGUAGES, LOCALES } from '~constants/index'

import styles from './LangSelector.module.css'

type Props = {
  showAbove?: boolean
  gray?: boolean
}

const LangSelector = ({ showAbove = false, gray = false }: Props) => {
  const [
    { locale: currentLocale },
    setLocaleState,
    { isLoading: isLanguageLoading, data: localeData }
  ] = useLocalization()

  const isDarkTheme = useThemeState()
  const langSelectorRef = useRef(null)
  const menuRef = useRef(null)
  const [isOpen, setOpen] = useState(false)

  useOutsideClick(langSelectorRef, () => setOpen(false))

  const langSelectorClass = isDarkTheme
    ? styles.langSelectorDark
    : styles.langSelector
  const grayClass = gray ? styles.gray : ''

  const triangleClass = showAbove
    ? isOpen
      ? styles.triangle
      : styles.triangleReverse
    : isOpen
      ? styles.triangleReverse
      : styles.triangle

  return (
    <div ref={langSelectorRef} className={`${langSelectorClass} ${grayClass}`}>
      <button
        onClick={() => setOpen((isOpen) => !isOpen)}
        className={styles.langSelectorInner}
        type="button"
        disabled={isLanguageLoading}
        aria-label={localeData?.switchLanguage.switchLang ?? 'Switch language'}
        aria-expanded={isOpen}
        aria-haspopup
        data-testid="langSelector"
      >
        <LanguageIcon width={20} height={20} className={styles.icon} />
        <span>{LANGUAGES.get(currentLocale)}</span>
        <div className={triangleClass} />
      </button>

      {isOpen && (
        <ul
          ref={menuRef}
          className={showAbove ? styles.dropDownAbove : styles.dropDownBelow}
          role="menu"
        >
          {LOCALES.map((locale) => (
            <li role="menuitem" key={locale}>
              <button
                type="button"
                disabled={locale === currentLocale}
                onClick={() => {
                  setLocaleState(locale as 'en' | 'ua')
                  setOpen(false)
                }}
              >
                {LANGUAGES.get(locale)}
              </button>
            </li>
          ))}
        </ul>
      )}
    </div>
  )
}

export default LangSelector
