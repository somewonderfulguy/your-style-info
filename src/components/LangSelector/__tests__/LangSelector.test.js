import React from 'react'
import {render, fireEvent} from '@testing-library/react'

import * as mockThemeContext from 'contexts/themeContext'
import LangSelector from '..'

jest.spyOn(mockThemeContext, 'useTheme').mockImplementation(() => ({isDarkTheme: false}))
afterEach(() => jest.clearAllMocks())

expect.addSnapshotSerializer({
  test(val) {return true},
  print(val) {return val.replace(/\n\+\s{0,}<ul(\n.*){1,}/gm, '')}
})

const setup = (showAbove, gray) => (
  render(<LangSelector showAbove={showAbove} gray={gray} />)
)

test('snapshot diff: dropdown above/below, default/gray colour, light/dark mode', () => {
  mockThemeContext.useTheme.mockReturnValueOnce({isDarkTheme: false})
  const componentDefault = setup().asFragment()

  mockThemeContext.useTheme.mockReturnValueOnce({isDarkTheme: true})
  const componentWithProps = setup(true, true).asFragment()

  expect(mockThemeContext.useTheme).toHaveBeenCalledTimes(2)
  expect(componentDefault).toMatchDiffSnapshot(componentWithProps, {
    contextLines: 1,
    aAnnotation: 'default props and theme',
    bAnnotation: 'drop-up, gray colour, dark mode'
  })
})

// TODO update prop! and combine two test in one
test('snapshot diff: aria state and css-class of triangle when open/hidden menu', () => {
  const { getByLabelText, asFragment } = setup()
  const beforeClick = asFragment()
  fireEvent.click(getByLabelText('Switch language'))

  expect(beforeClick).toMatchDiffSnapshot(asFragment(), {
    contextLines: 1,
    aAnnotation: 'collapsed menu',
    bAnnotation: 'expanded menu'
  })
})

test('snapshot diff: css-class of triangle when open/hide menu, preset: showing menu above the language selector (drop-up)', () => {
  const { getByLabelText, asFragment } = setup(true)
  const beforeClick = asFragment()
  fireEvent.click(getByLabelText('Switch language'))

  expect(beforeClick).toMatchDiffSnapshot(asFragment(), {
    contextLines: 1,
    aAnnotation: 'collapsed menu',
    bAnnotation: 'expanded menu'
  })
})

test('should show/hide menu when clicking on language selector', () => {
  const { getByLabelText, getByRole, queryByRole } = setup()
  const langSelectorBtn = getByLabelText('Switch language')

  expect(queryByRole('list')).toBeNull()

  fireEvent.click(langSelectorBtn)
  expect(getByRole('list')).toBeInTheDocument()

  fireEvent.click(langSelectorBtn)
  expect(queryByRole('list')).toBeNull()
})

test('should hide menu after selecting a language', () => {
  const { getByRole, queryByRole, getByLabelText } = setup()

  fireEvent.click(getByLabelText('Switch language'))
  fireEvent.click(getByRole('list').querySelector('button:not(:disabled)'))

  expect(queryByRole('list')).toBeNull()
})

test('should hide menu if clicked somewhere outside', () => {
  const { queryByRole, getByLabelText } = setup()

  fireEvent.click(getByLabelText('Switch language'))
  fireEvent.click(document)

  expect(queryByRole('list')).toBeNull()
})