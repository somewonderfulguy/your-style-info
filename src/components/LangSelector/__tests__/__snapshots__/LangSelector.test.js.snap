// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot diff: aria state and css-class of triangle when open/hidden menu 1`] = `
Snapshot Diff:
- collapsed menu
+ expanded menu

@@ -5,3 +5,3 @@
      <button
-       aria-expanded="false"
+       aria-expanded="true"
        aria-label="Switch language"
@@ -24,5 +24,26 @@
        <div
-         class="triangle"
+         class="triangleReverse"
        />
      </button>
`;

exports[`snapshot diff: css-class of triangle when open/hide menu, preset: showing menu above the language selector (drop-up) 1`] = `
Snapshot Diff:
- collapsed menu
+ expanded menu

@@ -5,3 +5,3 @@
      <button
-       aria-expanded="false"
+       aria-expanded="true"
        aria-label="Switch language"
@@ -24,5 +24,26 @@
        <div
-         class="triangleReverse"
+         class="triangle"
        />
      </button>
`;

exports[`snapshot diff: dropdown above/below, default/gray colour, light/dark mode 1`] = `
Snapshot Diff:
- default props and theme
+ drop-up, gray colour, dark mode

@@ -2,3 +2,3 @@
    <div
-     class="langSelector "
+     class="langSelectorDark gray"
    >
@@ -24,3 +24,3 @@
        <div
-         class="triangle"
+         class="triangleReverse"
        />
`;
